optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/common/AbstractAccumulateFunction.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.score.stream.drools.common;
+
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.io.Serializable;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import org.kie.api.runtime.rule.AccumulateFunction;
+import org.optaplanner.core.api.score.stream.uni.UniConstraintCollector;
+
+/**
+ * Adapts a CS constraint collector to {@link AccumulateFunction}.
+ *
+ * <p>
+ * {@link AccumulateFunction} has two behaviors:
+ *
+ * <ul>
+ * <li>First, using the {@link #accumulate(Object, Object)} method, it returns the accumulated object which is
+ * later consumed by {@link #reverse(AccumulateContext, Object)}.</li>
```

Comment:
Is it later consumed by `reverse`? This method throws `UnsupportedOperationException`.